### PR TITLE
[IMP] base_action_rule : Convert to new api

### PR DESCRIPTION
**Pull request by ara-odoo**
Task:https://www.odoo.com/web#id=12030&view_type=form&model=project.task&action=327
Pad: https://pad.odoo.com/p/r.j21wAI25Ei3COXst

### CODE REVIEW

File: addons/base_action_rule/models/base_action_rule.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        elif self.kind in ['on_write', 'on_create_or_write']:
+            self.trg_date_id = False
+            self.trg_date_range = False
+            self.trg_date_range_type = False
+        elif self.kind == 'on_time':
+            self.filter_pre_id = False
+
+    @api.onchange('filter_pre_id')
+    def _onchange_filter_pre_id(self):
+        self.filter_pre_domain = self.env['ir.filters'].browse(self.filter_pre_id.id).domain
```
Comment (ara-odoo) [inline addons/base_action_rule/models/base_action_rule.py]: no need to browse, you can directly use self.filter_pre_id.domain

File: addons/base_action_rule/models/base_action_rule.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+
+    @api.onchange('filter_id')
+    def _onchange_filter_id(self):
+        self.filter_domain = self.env['ir.filters'].browse(self.filter_id.id).domain
```
Comment (ara-odoo) [inline addons/base_action_rule/models/base_action_rule.py]: same comment

File: addons/base_action_rule/models/base_action_rule.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+    @api.multi
+    def unlink(self):
+        res = super(BaseActionRule, self).unlink()
+        self._update_cron()
+        return res
+
+    @api.onchange('model_id')
+    def _onchange_model_id(self):
+        if self.model_id:
+            self.model = self.env['ir.model'].browse(self.model_id.id).model
```
Comment (ara-odoo) [inline addons/base_action_rule/models/base_action_rule.py]: same comment

File: addons/base_action_rule/models/base_action_rule.py (hunk 1 above)
Comment (htr-odoo) [inline addons/base_action_rule/models/base_action_rule.py]: Done.

File: addons/base_action_rule/models/base_action_rule.py (hunk 2 above)
Comment (htr-odoo) [inline addons/base_action_rule/models/base_action_rule.py]: Done.

File: addons/base_action_rule/models/base_action_rule.py (hunk 3 above)
Comment (htr-odoo) [inline addons/base_action_rule/models/base_action_rule.py]: Done.

File: addons/base_action_rule/models/base_action_rule.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+        'ir.filters', string='Before Update Filter',
+        ondelete='restrict', domain="[('model_id', '=', model_id.model)]",
+        help="If present, this condition must be satisfied before the update of the record.")
+    filter_pre_domain = fields.Char(string='Before Update Domain', help="If present, this condition must be satisfied before the update of the record.")
+    filter_id = fields.Many2one(
+        'ir.filters', string='Filter',
+        ondelete='restrict', domain="[('model_id', '=', model_id.model)]",
+        help="If present, this condition must be satisfied before executing the action rule.")
+    filter_domain = fields.Char(string='Domain', help="If present, this condition must be satisfied before executing the action rule.")
+    last_run = fields.Datetime(string='Last Run', readonly=True, copy=False)
```
Comment (ara-odoo) [inline addons/base_action_rule/models/base_action_rule.py]: You can remove the string parameter here, as the ORM will generate a capitalized version of the field name by default

File: addons/base_action_rule/models/base_action_rule.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+
+    @api.multi
+    def unlink(self):
+        res = super(BaseActionRule, self).unlink()
+        self._update_cron()
+        return res
+
+    @api.onchange('model_id')
+    def _onchange_model_id(self):
+        if self.model_id:
```
Comment (ara-odoo) [inline addons/base_action_rule/models/base_action_rule.py]: no need to check `if self.model_id`

File: addons/base_action_rule/models/base_action_rule.py (hunk 4 above)
Comment (htr-odoo) [inline addons/base_action_rule/models/base_action_rule.py]: Done.

File: addons/base_action_rule/models/base_action_rule.py (hunk 5 above)
Comment (htr-odoo) [inline addons/base_action_rule/models/base_action_rule.py]: Done.

**Comment by hmo-odoo** [comment]
continue on PR: https://github.com/odoo/odoo/pull/5837

File: addons/base_action_rule/models/base_action_rule.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+        # Note: the patched methods create and write must be defined inside
+        # another function, otherwise their closure may be wrong. For instance,
+        # the function create refers to the outer variable 'create', which you
+        # expect to be bound to create itself. But that expectation is wrong if
+        # create is defined inside a loop; in that case, the variable 'create'
+        # is bound to the last function defined by the loop.
+        #
+        def make_create():
+            """ instanciate a create method that processes action rules """
+            def create(self, cr, uid, vals, context=None, **kwargs):
```
Comment (dreispt) [inline addons/base_action_rule/models/base_action_rule.py]: Shouldn't this be decorated with `@api.model`?

File: addons/base_action_rule/models/base_action_rule.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+                    post_ids = action_model._filter(cr, uid, action, action.filter_id, pre_ids[action], domain=action.filter_domain, context=context)
+                    if post_ids:
+                        action_model._process(cr, uid, action, post_ids, context=context)
+                return True
+            return write
+
+        updated = False
+        if ids is None:
+            ids = self.search(cr, SUPERUSER_ID, [])
+        for action_rule in self.browse(cr, SUPERUSER_ID, ids):
```
Comment (dreispt) [inline addons/base_action_rule/models/base_action_rule.py]: Browse shouldn't be needed, the search now return a recordset.

File: addons/base_action_rule/models/base_action_rule.py [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+            if model_obj and not hasattr(model_obj, 'base_action_ruled'):
+                # monkey-patch methods create and write
+                model_obj._patch_method('create', make_create())
+                model_obj._patch_method('write', make_write())
+                model_obj.base_action_ruled = True
+                updated = True
+        return updated
+
+    @api.v8
+    def _register_hook(self):
```
Comment (dreispt) [inline addons/base_action_rule/models/base_action_rule.py]: Are we defining another method with the same name, but a different signature?
Why is this?

**Comment by hmo-odoo** [comment]
base_action_rule is already migrated . see PR: https://github.com/odoo/odoo/pull/8940

**Comment by dreispt** [comment]
Thanks!

**Pull request closed without merging**